### ISSUE TITLE
sst takes longer than needed buffer from bp3

### ISSUE THREAD

**Issue by germasch** [ornladios/ADIOS2#1320]
@eisenhauer, I noticed that SstWriter uses `m_Data.m_Buffer.size()` to get the size of the provided buffer. That actually returns the capacity of `m_Data`, though, that is, it's potentially larger than the actual serialized data. It's not likely to have any actual consequences (I don't know if that size is even used at all. If the buffer were copied, it could potentially lead to copying more data than needed, but I don't thing it is ever copied in the first place.) 

Potential fix:

```diff
diff --git a/source/adios2/engine/sst/SstWriter.cpp b/source/adios2/engine/sst/SstWriter.cpp
index b17f7f92..e0e971a2 100644
--- a/source/adios2/engine/sst/SstWriter.cpp
+++ b/source/adios2/engine/sst/SstWriter.cpp
@@ -161,7 +161,7 @@ void SstWriter::EndStep()
         newblock->metadata.DataSize =
             m_BP3Serializer->m_Metadata.m_Buffer.size();
         newblock->metadata.block = m_BP3Serializer->m_Metadata.m_Buffer.data();
-        newblock->data.DataSize = m_BP3Serializer->m_Data.m_Buffer.size();
+        newblock->data.DataSize = m_BP3Serializer->m_Data.m_Position;
         newblock->data.block = m_BP3Serializer->m_Data.m_Buffer.data();
         newblock->serializer = m_BP3Serializer;
         TAU_STOP("Marshaling overhead");
```

That's from how BP3Writer does it:
```cxx
void BP3Writer::WriteData(const bool isFinal, const int transportIndex)
{
    size_t dataSize = m_BP3Serializer.m_Data.m_Position;

...
```

**Comment by eisenhauer** [comment]
@JasonRuonanWang, kicking to you because it's in the BP3 marshaling code...

**Comment by JasonRuonanWang** [comment]
I don't think std::vector<>::size() returns the capacity. It returns the size literally. Instead, std::vector<>::capacity() returns the capacity. 

In any way, at this point, m_Buffer.size() should be exactly the data size that has been used. It should never be over allocated. If it is then I would consider it a bug. So I don't see there is any problem with this.

**Comment by germasch** [comment]
This piece of code is a bit complicated, as there is a different between BufferSTL and std::vector.  You are right, of course, about std::vector's size() and capacity(). However, BufferSTL is essentially a buffer implemented on top of std::vector, ie., the underlying data storage is provided by the m_Buffer, but BufferSTL adds its own features on top, in particular, it keeps track of the position in the buffer (m_Position) and also the absolutePosition in the resulting file (m_AbsolutePosition). The underlying m_Buffer starts out at 16kb by default, but of course it's actually empty, indicated by m_Position = 0. As things are serialized into the buffer, m_Position is increased.  If running out of space, Resize() will resize the underlying m_Buffer by some exponential growth factor (e.g., from 16kb -> 24kb), but the size of the BufferSTL is still smaller, given by m_Position.

In other words, BufferSTL mostly behaves as std::vector, but:
* BufferSTL's size (how many bytes have been put into it) is given by m_Position, not m_Buffer.size()
* BufferSTL's capacity (how many bytes it can fit before it's necessary to realloc it) is given by m_Buffer.size() (which should also be equal to m_Buffer.capacity() because of how the Resize function is implemented.)

You can try it yourself, just put some print statements in there and then just write a little bit of data in a step. m_Buffer.size() will be 16kB because that is it's default initial size, while the actual buffer size (m_Position) will be much smaller.

**Comment by JasonRuonanWang** [comment]
Okay, if that's the case then we can change it, as long as it passes all tests. 

But I doubt it will affect anything, because this piece of code only gives the pointer and size for SST to temporarily hold. The actual amount of data being transferred is not depending on this, but rather how the readers issue the SST remote read call. As long as the readers get the correct start and count information in the metadata, they should only request the pieces they require, regardless the size that is registered here. 

But again, if it passes all tests then I am good either way.

**Comment by eisenhauer** [comment]
Let me put in a PR.  True, this overestimate wouldn't make any (or at least not much) difference in the current mode of reading data in BP marshaling, but the current mode maybe isn't going to be the last word here, and having the right size is important.  (In RDMA mode, we would be mapping based on this size, and the overconsumption of RDMA resources isn't great.)

**Comment by JasonRuonanWang** [comment]
@eisenhauer Sure, if you think that's better then please go ahead. I am not an expert in either SST or BP, and actually there are a lot of things which I don't understand in either of them. So I have to be conservative in changes which do not seem absolutely necessary. Otherwise whenever the changes cause extra problems I have to coordinate multiple people to solve

**Comment by eisenhauer** [comment]
PR #1325 has been merged.
